_posts/2020-07-30-axon-kompleksowe-testowanie-aplikacji.md

```diff
@@ -0,0 +1,386 @@
+```java
+@ActiveProfiles("test")
+public class TestContainers {
+
+    private static final int MONGO_PORT = 29019;
+    private static final int AXON_HTTP_PORT = 8024;
+    private static final int AXON_GRPC_PORT = 8124;
+
+    public static void startAxonServer() {
+        GenericContainer axonServer = new GenericContainer("axoniq/axonserver:latest")
+                .withExposedPorts(AXON_HTTP_PORT, AXON_GRPC_PORT)
+                .waitingFor(
+                        Wait.forLogMessage(".*Started AxonServer.*", 1)
+                );
+        axonServer.start();
+
+        System.setProperty(
+                "ENV_AXON_GRPC_PORT", 
+                String.valueOf(axonServer.getMappedPort(AXON_GRPC_PORT)));
+    }
+
+    public static void startMongo() {
+        GenericContainer mongo = new GenericContainer("mongo:latest")
+                .withExposedPorts(MONGO_PORT)
+                .withEnv("MONGO_INITDB_DATABASE", "moviekeeper")
+                .withCommand(String.format("mongod --port %d", MONGO_PORT))
+                .waitingFor(
+                        Wait.forLogMessage(".*waiting for connections.*", 1)
+                );
+        mongo.start();
+
+        System.setProperty(
+                "ENV_MONGO_PORT", 
+                String.valueOf(mongo.getMappedPort(MONGO_PORT)));
+    }
+}
+```
+Należy tu jednak pamiętać o tym, że **withExposedPorts** wystawia porty tylko **wewnątrz kontenera**, potrzebny więc był sposób na pozyskanie portów, do których testy będą mogły się połączyć.
+Testcontainers przy każdym restarcie kontenera wystawia go na losowym wolnym porcie z danego zakresu, istnieje jednak metoda na pobranie tych portów w runtime.
+Robię to w ostatniej instrukcji, każdej z metod jednocześnie wrzucając znalezione wartości do zmiennych środowiskowych **ENV_AXON_GRPC_PORT** oraz **ENV_MONGO_PORT**.
+Zmienne te używam w yamlu konfiguracyjnym pod testy:
+```yaml
+spring:
+  data:
+        mongodb:
+            uri: mongodb://localhost:${ENV_MONGO_PORT}/moviekeeper
+axon:
+  axonserver:
+        servers: localhost:${ENV_AXON_GRPC_PORT}
+```
+Metody **startMongo** i **startAxonServer** wykorzystałem w klasie, z której dziedziczą wszystkie testy integracyjne:
+```java
+@ExtendWith(SpringExtension.class)
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ActiveProfiles("test")
+public class CommonIntegrationSetup {
+    ...
+    @BeforeAll
+    public static void beforeAll() {
+        TestContainers.startAxonServer();
+        TestContainers.startMongo();
+    }
+    ...
+}
+```
+Należy pamiętać o **ustawieniu profilu** i **pliku konfiguracyjnym** pod ten profil - zdarzyło mi się puścić testy (bez tych dwóch rzeczy skonfigurowanych), podczas gdy aplikacja chodziła "produkcyjnie" - można łatwo zgadnąć, do jakiego AxonServera owe testy się podłączyły. :)
+
+## Skonfigurowane. Do dzieła!
+W mojej aplikacji w momencie, gdy uda się znaleźć film o żądanym tytule dzieją się dwie operacje:
```

Comment:
uda się daje poczucie losowości. Może więc:
W mojej aplikacji w momencie, gdy uda się znaleźć film o --> W mojej aplikacji w momencie, gdy znajdę film o

Reply:
Druga kwestia- to nie są operacje. To są czynności/zdarzenia/kroki. Chyba że zmienisz formę ich zapisania na:
1. Zwracanie szczegółów znalezionego filmu.
2. Wysyłanie command'a, który mówi **znajdź trailery i obsadę dla tego filmu

Reply:
Done